examples/pytorch/rgcn/entity_classify_mp.py

```diff
@@ -269,54 +326,64 @@ def run(proc_id, n_gpus, args, devices, dataset):
 
             forward_time.append(t1 - t0)
             backward_time.append(t2 - t1)
-            print("Epoch {:05d}:{:05d} | Train Forward Time(s) {:.4f} | Backward Time(s) {:.4f}".
-                   format(epoch, i, forward_time[-1], backward_time[-1]))
             train_acc = th.sum(logits.argmax(dim=1) == labels[seeds]).item() / len(seeds)
-            print("Train Accuracy: {:.4f} | Train Loss: {:.4f}".
-                  format(train_acc, loss.item()))
-
-        # only process 0 will do the evaluation
-        if proc_id == 0:
-            model.eval()
-            eval_logtis = []
-            eval_seeds = []
-            for i, sample_data in enumerate(val_loader):
-                seeds, blocks = sample_data
-                feats = embed_layer(blocks[0].srcdata[dgl.NID].to(dev_id),
-                                    blocks[0].srcdata[dgl.NTYPE].to(dev_id),
-                                    node_feats)
-                logits = model(blocks, feats)
-                eval_logtis.append(logits)
-                eval_seeds.append(seeds)
-            eval_logtis = th.cat(eval_logtis)
-            eval_seeds = th.cat(eval_seeds)
-            val_loss = F.cross_entropy(eval_logtis, labels[eval_seeds])
-            val_acc = th.sum(eval_logtis.argmax(dim=1) == labels[eval_seeds]).item() / len(eval_seeds)
-            print("Validation Accuracy: {:.4f} | Validation loss: {:.4f}".
-                    format(val_acc, val_loss.item()))
+            if i % 100 and proc_id == 0:
+                print("Train Accuracy: {:.4f} | Train Loss: {:.4f}".
+                    format(train_acc, loss.item()))
+        print("Epoch {:05d}:{:05d} | Train Forward Time(s) {:.4f} | Backward Time(s) {:.4f}".
+            format(epoch, i, forward_time[-1], backward_time[-1]))
+
+        if (queue is not None) or (proc_id == 0):
+            val_logits, val_seeds = evaluate(model, embed_layer, val_loader, node_feats)
+            if queue is not None:
+                queue.put((val_logits, val_seeds))
+
+            # gather evaluation result from multiple processes
+            if proc_id == 0:
+                if queue is not None:
+                    val_logits = []
+                    val_seeds = []
+                    for i in range(n_gpus):
+                        log = queue.get()
+                        val_l, val_s = log
+                        val_logits.append(val_l)
+                        val_seeds.append(val_s)
+                    val_logits = th.cat(val_logits)
+                    val_seeds = th.cat(val_seeds)
+                val_loss = F.cross_entropy(val_logits, labels[val_seeds].cpu()).item()
+                val_acc = th.sum(val_logits.argmax(dim=1) == labels[val_seeds].cpu()).item() / len(val_seeds)
```

Comment:
this is a very common function for multiprocessing training and distributed training. we need to formalize it and provide an API for this.